nhs/data/allocation.py

```diff
@@ -0,0 +1,286 @@
+    ... ).collect()
+        shape: (3, 4)
+        ┌───────────┬──────────┬───────────┬──────────┐
+        │ CODE_2021 ┆ feature1 ┆ longitude ┆ latitude │
+        │ ---       ┆ ---      ┆ ---       ┆ ---      │
+        │ i64       ┆ i64      ┆ f64       ┆ f64      │
+        ╞═══════════╪══════════╪═══════════╪══════════╡
+        │ 123       ┆ 24       ┆ 134.56    ┆ -34.56   │
+        │ 45        ┆ 65       ┆ 456.12    ┆ -23.12   │
+        │ 3         ┆ 234      ┆ 21.124    ┆ 12.124   │
+        └───────────┴──────────┴───────────┴──────────┘
+    """
+    census_type_casted = census.with_columns(
+        pl.col(left_code_col).cast(coords.collect_schema()[right_code_col])
+    )
+
+    return census_type_casted.join(
+        coords, left_on=left_code_col, right_on=right_code_col, how="inner"
+    )
+
+
+@log_entry_exit()
+def sample_census_feature(
+    census: pl.LazyFrame, code_col: str, long_col: str, lat_col: str, feature_col: str
+):
+    """
+    Randomly sample rows from each group in the census LazyFrame using value of `feature_col` as sample size.
+
+    Parameters
+    ----------
+    census: pl.LazyFrame
+        A `LazyFrame` to sample from, with columns `code_col`, `long_col`, `lat_col`,
+        and `feature_col`.
+    code_col : str
+        Column in `census` that contains some identifier/code.
+    long_col : str
+        Column in `census` that contains the longitude.
+    lat_col : str
+        Column in `census` that contains the latitude.
+    feature_col : str
+        Column in `census` for a feature where each row is the
+        sample size for each group.
+
+    Returns
+    -------
+    pl.LazyFrame
+        A `LazyFrame` with the columns `[code_col, long_col, lat_col, feature_col]`
+        where each `code_col` group have N number of randomly sampled rows. N
+        is the value of `feature_col` for each group.
+
+    Examples
+    --------
+    >>> census = pl.LazyFrame({
+    ...     "code_col": ["A", "A", "A", "B"],
+    ...     "long_col": [1.0, 2.0, 3.0, 4.0],
+    ...     "lat_col": [10.0, 20.0, 30.0, 40.0],
+    ...     "feature_col": [4, 4, 4, 2],
+    ... })
+    >>> sample_census_feature(
+    ...     census, "code_col", "long_col", "lat_col", "feature_col"
+    ... ).collect()
+    shape: (6, 4)
+    ┌──────────┬─────────────┬──────────┬─────────┐
+    │ code_col ┆ feature_col ┆ long_col ┆ lat_col │
+    │ ---      ┆ ---         ┆ ---      ┆ ---     │
+    │ str      ┆ bool        ┆ f64      ┆ f64     │
+    ╞══════════╪═════════════╪══════════╪═════════╡
+    │ A        ┆ true        ┆ 1.0      ┆ 10.0    │
+    │ A        ┆ true        ┆ 1.0      ┆ 10.0    │
+    │ A        ┆ true        ┆ 2.0      ┆ 20.0    │
+    │ A        ┆ true        ┆ 3.0      ┆ 30.0    │
+    │ B        ┆ true        ┆ 4.0      ┆ 40.0    │
+    │ B        ┆ true        ┆ 4.0      ┆ 40.0    │
+    └──────────┴─────────────┴──────────┴─────────┘
```

Comment:
Shouldnt this sample output have shape (14, 2)? from [4 + 4 + 4 + 2] samples randomly assigned to any of the 4 locations.

Reply:
The example is correct, rows in `feature_col` are duplicated for every coordinate.

Basically, for SA1 region `"A"`, we have 4 people with the feature `feature_col`. The row is duplicated because we have different coordinates. Think of it as: we have this table

| code | coordinates | feature |
| --- | --- | --- |
| A | [(1, 1), (1, 2), (2, 3)] | 4 |

But instead of representing the coordinates as a list, we "explode" it into separate rows

| code | longitude | latitude | feature |
| --- | --- | --- | --- |
| A | 1 | 1 | 4 |
| A | 1 | 2 | 4 |
| A | 2 | 3 | 4 |